cpp/src/arrow/array/concatenate_test.cc

```diff
@@ -225,6 +225,86 @@ TEST_F(ConcatenateTest, DictionaryType) {
   });
 }
 
+TEST_F(ConcatenateTest, DictionaryTypeDifferentDictionaries) {
+  auto dict_type = dictionary(uint8(), utf8());
```

Comment:
Hmm... can you test with index types larger than 1 byte? There may be an issue with them in the current impl (not sure).

Reply:
I added a test and there was an issue.  Both the missing reinterpret cast and the way I was computing size (# bytes vs. # elements)